fuse_core/include/fuse_core/message_buffer_impl.h

```diff
@@ -137,14 +137,32 @@ void MessageBuffer<Message>::purgeHistory()
   {
     return;
   }
-  // Continue to remove the first entry from the buffer until we:
-  // (a) are left with only two entries, OR
-  // (b) the time delta between the beginning and end is within the buffer_length_
+
+  // Compute the expiration time carefully, as ROS can't handle negative times
   const ros::Time& ending_stamp = buffer_.back().first;
-  while ((buffer_.size() > 2)
-      && ((ending_stamp - buffer_.front().first) > buffer_length_))
+  ros::Time expiration_time;
+  if (ending_stamp.toSec() < buffer_length_.toSec())
+  {
+    expiration_time = ros::Time(0);
+  }
+  else
+  {
+    expiration_time = ending_stamp - buffer_length_;
+  }
+  // Remove buffer elements before the expiration time.
+  // Be careful to ensure that:
+  //  - at least two entries remains at all times
+  //  - the buffer covers *at least* until the expiration time. Longer is acceptable.
+  auto is_greater = [](const ros::Time& stamp, const typename Buffer::value_type& element) -> bool
```

Comment:
I believe with C++14, you can use `auto` for the arguments.

```suggestion
  auto is_greater = [](const auto& stamp, const auto& element) -> bool
```